cpp/src/arrow/types/struct.cc

```diff
@@ -17,4 +17,94 @@
 
 #include "arrow/types/struct.h"
 
-namespace arrow {}  // namespace arrow
+#include <sstream>
+
+namespace arrow {
+
+bool StructArray::EqualsExact(const StructArray& other) const {
+  if (this == &other) { return true; }
+  if (null_count_ != other.null_count_) { return false; }
+
+  bool equal_null_bitmap = true;
+  if (null_count_ > 0) {
+    equal_null_bitmap =
+        null_bitmap_->Equals(*other.null_bitmap_, util::bytes_for_bits(length_));
+  }
+
+  if (!equal_null_bitmap) { return false; }
+  bool values_equal = false;
```

Comment:
i'm pretty sure this code is incorrect.  You end up only checking if the last field array is equal.  You probably need to early terminate.